src/unix.rs

```diff
@@ -22,6 +24,89 @@ pub fn table_load(_opt: &Opt) -> Result<SMBiosData, Error> {
     table_load_from_dev_mem()
 }
 
+/// Load from /sys/firmware/dmi/tables/DMI
+fn table_load_from_sysfs() -> Result<SMBiosData, Error> {
+    let mut output = String::new();
+
+    writeln!(&mut output, "Getting SMBIOS data from sysfs.").unwrap();
+
+    let version: SMBiosVersion;
+    let entry_path = std::path::Path::new(SYS_ENTRY_FILE);
+
+    match SMBiosEntryPoint64::try_load_from_file(entry_path) {
+        Ok(entry_point) => {
+            version = SMBiosVersion {
+                major: entry_point.major_version(),
+                minor: entry_point.minor_version(),
+                revision: entry_point.docrev(),
+            };
+
+            writeln!(
+                &mut output,
+                "SMBIOS {}.{}.{} present.",
+                entry_point.major_version(),
+                entry_point.minor_version(),
+                entry_point.docrev()
+            )
+            .unwrap();
+
+            writeln!(
+                &mut output,
+                "Occupying {} bytes maximum.",
+                entry_point.structure_table_maximum_size()
+            )
+            .unwrap();
+
+            writeln!(
+                &mut output,
+                "Table at: {:#010X}.",
+                entry_point.structure_table_address()
+            )
+            .unwrap();
+        }
+        Err(err) => match err.kind() {
+            ErrorKind::InvalidData => match SMBiosEntryPoint32::try_load_from_file(entry_path) {
+                Ok(entry_point) => {
+                    version = SMBiosVersion {
+                        major: entry_point.major_version(),
+                        minor: entry_point.minor_version(),
+                        revision: 0,
+                    };
+
+                    writeln!(
+                        &mut output,
+                        "SMBIOS {}.{} present.",
+                        entry_point.major_version(),
+                        entry_point.minor_version()
+                    )
+                    .unwrap();
+
+                    writeln!(
+                        &mut output,
+                        "{} structures occupying {} bytes.",
+                        entry_point.number_of_smbios_structures(),
+                        entry_point.structure_table_length()
+                    )
+                    .unwrap();
+
+                    writeln!(
+                        &mut output,
+                        "Table at: {:#010X}.",
+                        entry_point.structure_table_address()
+                    )
+                    .unwrap();
+                }
+                Err(err) => return Err(err),
+            },
+            _ => return Err(err),
+        },
+    }
+
+    print!("{}", output);

```

Comment:
Looks good. The only question that I have is this `print!` statement. Not sure why this function prints to stdout, and not the one that uses this function `table_load_from_sysfs`

Reply:
Do you mean change the fn design from return `Result<SMBiosData, Error>` to `Result<(SMBiosData, String), Error>` and have the caller of the function do the print?

Reply:
Yep that worked well!  I love the capability of tuples, it's like a nameless struct.  The first example here shows using tuples for both in and out parameters: https://doc.rust-lang.org/rust-by-example/primitives/tuples.html